File: backend/src/controllers/auth.controller.ts
import { Request, Response } from 'express'
import { supabase } from '../config/supabase'

export const register = async (req: Request, res: Response) => {
  const { email, password, name, phone } = req.body
  // phone é opcional — usuário pode cadastrar sem telefone

  const { data, error } = await supabase.auth.signUp({
    email,
    password,
    options: {
      data: { name, phone } // esses dados ficam em raw_user_meta_data no Supabase
                            // o trigger handle_new_user lê daqui e salva no profiles
    }
  })

  if (error) return res.status(400).json({ error: error.message })
  return res.status(201).json({ user: data.user })
}

export const login = async (req: Request, res: Response) => {
  const { email, password } = req.body

  const { data, error } = await supabase.auth.signInWithPassword({
    email,
    password
  })

  if (error) return res.status(401).json({ error: error.message })
  return res.status(200).json({ user: data.user, session: data.session })
}

export const logout = async (req: Request, res: Response) => {
  const { error } = await supabase.auth.signOut()
  if (error) return res.status(400).json({ error: error.message })
  return res.status(200).json({ message: 'Logout realizado com sucesso' })
}

File: backend/internal/auth/auth_controller.go
package auth

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"beeplanerofc/backend-go/internal/supabaseauth"
)

type Controller struct {
	authClient *supabaseauth.Client
}

func NewController(authClient *supabaseauth.Client) *Controller {
	return &Controller{authClient: authClient}
}

type registerRequest struct {
	Email    string `json:"email" binding:"required"`
	Password string `json:"password" binding:"required"`
	Name     string `json:"name"`
	Phone    string `json:"phone"` // opcional — usuário pode cadastrar sem telefone
}

// @Summary      Cadastrar usuário
// @Tags         Auth
// @Accept       json
// @Produce      json
// @Param        body body registerRequest true "Dados de cadastro"
// @Success      201 {object} map[string]interface{}
// @Failure      400 {object} map[string]string
// @Router       /auth/register [post]
func (ctrl *Controller) Register(c *gin.Context) {
	var req registerRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Dados inválidos"})
		return
	}

	user, err := ctrl.authClient.SignUp(req.Email, req.Password, map[string]any{
		"name":  req.Name,
		"phone": req.Phone,
	})
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, gin.H{"user": user})
}

type loginRequest struct {
	Email    string `json:"email" binding:"required"`
	Password string `json:"password" binding:"required"`
}

// @Summary      Login
// @Tags         Auth
// @Accept       json
// @Produce      json
// @Param        body body loginRequest true "Credenciais"
// @Success      200 {object} map[string]interface{}
// @Failure      401 {object} map[string]string
// @Router       /auth/login [post]
func (ctrl *Controller) Login(c *gin.Context) {
	var req loginRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Dados inválidos"})
		return
	}

	user, session, err := ctrl.authClient.SignInWithPassword(req.Email, req.Password)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"user": user, "session": session})
}

// @Summary      Logout
// @Tags         Auth
// @Produce      json
// @Success      200 {object} map[string]string
// @Router       /auth/logout [post]
func (ctrl *Controller) Logout(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{"message": "Logout realizado com sucesso"})
}
